### ISSUE TITLE
render_expression drops a boolean mask: mean(W[W > 0]) becomes mean(W)

### ISSUE THREAD

**Issue by leon-k-martin** [virtual-twin/tvbo#86]
`render_expression` silently drops a boolean mask from a transform equation, so a recipe declaring a masked reduction gets an unmasked one.

## Reproduce

```python
from tvbo import Network
import numpy as np
W = np.array([[0, 2.0, 0], [4.0, 0, 3.0], [0, 5.0, 0]])
n = Network.from_matrix(weights=W, lengths=np.zeros_like(W))
n.add_transform("weight", "W / mean(W[W > 0])")
# emitted / evaluated expression is:  W/jnp.mean(W)
```

The `[W > 0]` index is gone. `mean(W)` over a sparse connectome includes every zero entry, so the divisor is far smaller than the mean non-zero weight — for the matrix above, 1.56 instead of 3.5.

## Where it bites

`tvbo/database/experiments/Delay_Speed_Synchronization.yaml:55` declares

```yaml
rhs: "W / mean(W[W > 0])"   # normalize so the mean non-zero weight is 1 (keeps G O(1))
```

so the recipe's stated intent — and the reason `G` is expected to stay O(1) — is not what runs.

## Why nothing fails

The runtime (`Network._apply_transform`) and the tvboptim emitter both go through `render_expression`, so they agree with each other and every byte-identity test passes. The divergence is between the *declaration* and *both* implementations, which no test compares.

## Scope

The loss is in the sympy parse/print layer (`tvbo/parse/expression.py` / `tvbo/codegen/code.py`), not in the transform lowering — #85 reworked the latter and deliberately left this alone.

A fix should either honour the mask or reject it at parse time. Silently dropping an index from a declared equation is the worst of the three.

**Comment by leon-k-martin** [comment]
**Status update — triaged against `dev` at `636cb868`** (after #101, #98, #94, #88 merged).

**Verdict: still open — not fixed**

The claim is partial where it was stated as total ("the reported bug is closed"), and the part left undone is the part the issue actually names.

What the issue says, read in full (not summarized):
- Title: "render_expression drops a boolean mask: mean(W[W > 0]) becomes mean(W)". The defect is named against `render_expression`, not against the transform path.
- Scope section, verbatim: "The loss is in the sympy parse/print layer (`tvbo/parse/expression.py` / `tvbo/codegen/code.py`), **not in the transform lowering** — #85 reworked the latter and **deliberately left this alone**."
- Remedy, verbatim: "A fix should either honour the mask or reject it at parse time. Silently dropping an index from a declared equation is the worst of the three."
(There are no comments on the issue; state is `open`.)

The claimed fix is the third rework of the transform lowering — precisely the location the issue rules out — and it does not touch either file the issue names. `git log 636cb868^1..636cb868 -- tvbo/codegen/code.py tvbo/parse/expression.py` is EMPTY: the entire #94 merge modified neither. `git show --stat b660016b` and `153ce4b1` confirm the same.

I reproduced the issue's titular defect on dev myself, through tvbo's own general parse entry point:
  parse_eq("x / mean(x[x > 0])") -> render_expression(format="jax") == 'x/jnp.mean(x)'
That is the issue's headline, verbatim, unfixed. The claimant conceded this as a "latent hazard", but it is not latent — it is reachable through a public API with no transform involved:

  Function(name="masked_mean_rate", arguments=["r"], equation={"rhs": "mean(r[r > 0])"})
    .render_code(format="jax")   -> "def masked_mean_rate(r):\n    return jnp.mean(r)"      # mask gone
    .render_code(format="numpy") -> "def masked_mean_rate(r):\n    return np.mean(r[np.greater(r, 0)])"  # mask honoured

`tvbo.classes.function.Function` is the class observations, losses and derived quantities use, and `render_code(format="jax")` is its default path (function.py:123, :284). `tvbo/parse/expression.py:279-283` auto-promotes any subscripted name to `IndexedBase`, so this parses cleanly from ANY equation string in the datamodel — dynamics rhs, coupling, observation, loss, algorithm expression. Every one of the ~20 `format='jax'` render sites I grepped goes through the same unpatched printer. Only the network-transform consumer was rerouted around it.

Worse than before: the printers now disagree. On dev the same declared equation gives jax `jnp.mean(x)`, numpy `np.mean(x[np.greater(x, 0)])`, julia `mean(x[x > 0])`, python `statistics.mean(x[(x > 0)])`. The issue's own framing of why this is dangerous was "the divergence is between the declaration and *both* implementations, which no test compares." That divergence now also runs backend-to-backend, and no test compares that either. transforms.py's docstring even asserts "no two backends can disagree about what it means" — true only because the transform path routes around the broken printer, not because the printer was fixed.

No regression guard exists for the named defect. `grep -rln render_expression tests/` finds four test files, none of which passes a boolean-indexed expression; the only mask coverage in the tree is tests/test_weight_transform_codegen.py, which exercises the lowering, never the printer.

Neither of the issue's two acceptable remedies is met at the layer it names: `render_expression` neither honours the mask nor rejects it. The `ValueError` at transforms.py:164-171 is a transform-lowering guard, not a parse-time rejection, and it fires only on the one path that was already rerouted — so the "worst of the three" behaviour the issue calls out is still the live default everywhere else.

To be fair and precise about what IS real: the transform-path symptom is genuinely fixed and genuinely guarded. Both commits are ancestors of `dev` (verified with `git branch --contains`), `tvbo/codegen/transforms.py` exists at every cited line, `Delay_Speed_Synchronization.yaml:55` now honours its mask, and all 20 tests in tests/test_weight_transform_codegen.py pass locally (120s), including the two that assert against an independent numpy oracle `SPARSE[SPARSE>0].mean()`. So this is not a fix-on-a-branch problem and not a fabricated-evidence problem. It is a scope problem: the fix closes one consumer of a defect the issue filed against the shared printer, and the issue explicitly pre-empted exactly this by saying the transform lowering was not where the loss is. Closing #86 on this basis retires an open, reproducible, untested codegen-correctness bug that will be rediscovered the first time anyone writes a masked reduction in an observation or loss.

> The first pass recorded this as more fixed than it is. On re-check:
> - The issue's own Scope section rules out the layer that was fixed: 'The loss is in the sympy parse/print layer (tvbo/parse/expression.py / tvbo/codegen/code.py), not in the transform lowering — #85 reworked the latter and deliberately left this alone.' The #94 stack is a third rework of the transform lowering.
> - `cd /Users/leonmartin_bih/tools/tvbo && git log --oneline 636cb868^1..636cb868 -- tvbo/codegen/code.py tvbo/parse/expression.py` returns EMPTY — the entire merge touched neither file the issue names. Per-commit stats agree: b660016b touched schema/tvbo_datamodel.yaml, tests/test_weight_transform_codegen.py, tvbo/classes/network.py, tvbo/codegen/transforms.py, Delay_Speed_Synchronization.yaml only; 153ce4b1 likewise plus docs/templates/utils.
> - Live reproduction of the issue's TITLE on dev, via tvbo's own general parser: `parse_eq('x / mean(x[x > 0])')` then `render_expression(..., format='jax')` yields `'x/jnp.mean(x)'`. Also `mean(x[x>0])`->`jnp.mean(x)`, `sum(x[x>0])`->`jnp.sum(x)`, `max(x[x>0]) - min(x[x>0])`->`jnp.max(x) - jnp.min(x)`.
> - The bug is NOT latent — it is reachable through a public API with no transform involved. `tvbo.classes.function.Function(name='masked_mean_rate', arguments=['r'], equation={'rhs': 'mean(r[r > 0])'}).render_code(format='jax')` emits `def masked_mean_rate(r):\n    return jnp.mean(r)` while `format='numpy'` emits `np.mean(r[np.greater(r, 0)])`. Function is the class observations/losses/derived quantities use; function.py:123 and :284 default to format='jax'.

Evidence:
- Live repro on dev (HEAD 6f832f4c, 153ce4b1 confirmed an ancestor of local `dev`): Network.from_matrix(W, transforms=[{name: weight, equation: {rhs: 'weight / mean(weight[weight > 0])'}}]) emits ('weight*jnp.sum(jnp.where(_mask0, 1, 0))/jnp.sum(jnp.where(_mask0, weight, 0))', ['weight = weights', '_mask0 = jnp.greater(weight, 0)']) and the runtime matrix equals W / W[W>0].mean() (divisor 3.5, not the 1.556 the issue reports)
- tvbo/codegen/transforms.py:93 `canonical_reductions` rewrites both mask spellings to `red(operand, predicate)`; :107 `lower_reductions` lowers to `Piecewise` (mean -> kept-sum over kept-count); :143 `prepare` is the single entry point
- tvbo/codegen/transforms.py:164-171 — `prepare` raises ValueError naming 'outside a reduction' when a boolean-indexed node survives, instead of dropping it
- tvbo/classes/network.py:4617 and :4626 — `Network.transform_expression` imports and calls `prepare(...)`; the runtime consumes it at network.py:4684 (`exp, masks = self.transform_expression(func)`)
- tvbo/templates/tvboptim/utils.py:856 — the emitter (`weight_transform_codegen`, defined at utils.py:759) goes through the same `network.transform_expression(t)`, so declaration, runtime and kit share one lowering
- tvbo/database/experiments/Delay_Speed_Synchronization.yaml:55 — the recipe named in the issue now reads `rhs: "weight / mean(weight[weight > 0])"` and that mask is honoured

**PRs:** **#94 is merged but does not fix this.** This issue's Scope section excludes the transform-lowering layer, which is exactly what #94 reworked: `git log 636cb868^1..636cb868 -- tvbo/codegen/code.py tvbo/parse/expression.py` returns **0 commits**.

**Reproduced on `dev`:**
```
parse_eq("x / mean(x[x > 0])")
  jax    : x/jnp.mean(x)              <- mask dropped
  numpy  : x/np.mean(x[np.greater(x, 0)])
  julia  : x ./ mean(x[x > 0])
  python : x/statistics.mean(x[(x > 0)])
```
`jax` is the default format for `Function.render_code` (`tvbo/classes/function.py:123`), and no test passes a boolean-indexed expression through `render_expression`, so this is reachable from a public API and uncovered.


<sub>Triaged by re-running the issue against current `dev`. Every “already fixed” claim was independently re-checked by a second pass before being recorded here; three were overturned.</sub>
